drivers/mcp2515/mcp2515.c

```diff
@@ -0,0 +1,389 @@
+ */
+static void fill_extended_id(uint32_t id, uint8_t *bytebuf);
+
+int mcp2515_init(candev_mcp2515_t *dev, void(*irq_handler_cb)(void*))
+{
+    int res;
+    res = gpio_init_int(dev->conf->int_pin, GPIO_IN_PU, GPIO_FALLING, (gpio_cb_t)irq_handler_cb, (void *)dev);
+    if (res != 0) {
+        DEBUG("Error setting interrupt pin!\n");
+        return -1;
+    }
+    gpio_init(dev->conf->rst_pin, GPIO_OUT);
+
+    res = mcp2515_spi_init(dev);
+    if (res < 0){
+        return -1;
+    }
+
+    uint8_t cmd = MCP2515_RXB0CTRL_MODE_RECV_ALL;
+    res = mcp2515_spi_write(dev, MCP2515_RXB0CTRL , &cmd, 1);
+    if (res < 0) {
+        DEBUG("failed to set acceptance mode\n");
+        return -1;
+    }
+    return 0;
+}
+
+void mcp2515_reset(candev_mcp2515_t *dev)
+{
+    gpio_clear(dev->conf->rst_pin);
+    xtimer_usleep(RESET_DELAY);
+    gpio_set(dev->conf->rst_pin);
+    xtimer_usleep(ost_delay(dev));
+}
+
+static void fill_standard_id(uint32_t id, uint8_t *bytebuf)
+{
+    bytebuf[0] = (uint8_t) ((id & 0x000007F8UL) >> 3);  /* T/RXBnSIDH */
+    bytebuf[1] = (uint8_t) ((id & 0x00000007UL) << 5);  /* T/RXBnSIDL */
+    bytebuf[2] = (uint8_t) ((id & 0xFF000000UL) >> 24); /* T/RXBnEID8 */
+    bytebuf[3] = (uint8_t) ((id & 0x00FF0000UL) >> 16); /* T/RXBnEID0 */
+}
+
+static void fill_extended_id(uint32_t id, uint8_t *bytebuf)
+{
+    bytebuf[0] = (uint8_t) ((id & 0x1FE00000UL) >> 21); /* T/RXBnSIDH */
+    bytebuf[1] = (uint8_t) ((id & 0x001C0000UL) >> 13)
+        | (uint8_t) ((id & 0x00030000UL) >> 16) | 0x08; /* T/RXBnSIDL */
+    bytebuf[2] = (uint8_t) ((id & 0x0000FF00UL) >> 8);  /* T/RXBnEID8 */
+    bytebuf[3] = (uint8_t) (id & 0x000000FFUL);         /* T/RXBnEID0 */
+}
+
+int mcp2515_send(candev_mcp2515_t *dev, const struct can_frame *frame, int mailbox)
+{
+    uint8_t prio = 1; /* TODO: adjust priority */
+    uint8_t outbuf[BUFFER_SIZE];
+    uint8_t ctrl;
+    
+    struct can_frame framebuf;
+
+    if (frame->can_dlc > CAN_MAX_DLEN) {
+        return -1;
+    }
+
+    framebuf.can_id = frame->can_id;
+    framebuf.can_dlc = frame->can_dlc;
+    for(int i = 0; i < framebuf.can_dlc; i++) {
+        framebuf.data[i] = frame->data[i];
+    }
+
+    //TODO: for speedup, remove this check
+    mcp2515_spi_read(dev, TX_CTRL(mailbox), &ctrl, 1);
+    if (ctrl & MCP2515_TXBCTRL_TXREQ) {
+        DEBUG("Mailbox in use, TXB%dCTRL: 0x%02x\n", mailbox, ctrl);
+        return -1;
+    }
+
+    if(framebuf.can_id > CAN_SFF_MASK){
+        framebuf.can_id |= CAN_EFF_FLAG;
+    }
+
+    if ((framebuf.can_id & CAN_EFF_FLAG) == CAN_EFF_FLAG) {
+        fill_extended_id(framebuf.can_id, outbuf);
+    }
+    else {
+        fill_standard_id(framebuf.can_id, outbuf);
+    }
+
+    outbuf[4] = framebuf.can_dlc;
+    memcpy(outbuf+5, framebuf.data, framebuf.can_dlc);
+
+    /* set mailbox priority */
+    mcp2515_spi_write(dev, TX_CTRL(mailbox), &prio, 1);
+
+    mcp2515_spi_write_txbuf(dev, mailbox, outbuf, 5 + framebuf.can_dlc);
```

Comment:
32bit can_id + 8 bit can_dlc
I defined it as CAN_FIXED_LEN, because I couldn't think of a better name. I'm open for suggestions.

Reply:
I don't quite understand. There is the data length code (dlc) plus an offset of five. What is five, can we replace it?

Reply:
4 bytes can + 1 byte dlc = 5 bytes that are fixed. The remaining bytes (the payload) are variable. can_dlc determines the length of this payload. 

And yes, it can be replaced. I already did in yesterday's patch. (it doesn't show up in this code snippet because it refers to an older version). 

I probably did miss a few other places where this same define could be used though. I'll change those as well.